doc/manual/src/contributing/experimental-flags.md

```diff
@@ -0,0 +1,89 @@
+This section aims at describing the notion of a Nix "experimental
+flag", and how it fits into the big picture of the developpement of Nix.
+
+# What are these xp flags?
+
+Since Nix 2.4, some features of Nix can be flagged as
+experimental.Conceptually, this means that these features are not
+considered as part of Nix proper, but are only included in the code to
+make their developpement and testing easier.Practically, these require
+an explicit opt-in, by adding the name of the experimental feature to
+the `experimental-features` configuration option. This is really handy
+as it makes it harder to accidentally use an experimental feature
+without knowing it.
+
+# When should an experimental flag be used
+
+A change in the Nix codebase should be guarded by an experimental flag
+if it is likely to either introduce a regression of any kind, or if it
+changes the external interface of Nix (be it the language, the cli or
+anything else that's considered as "public").
+
+For example, these changes need to be guarded by an experimental flag:
+
+-   Any language change (new builtin, syntactic or semantic change,
+    etc..),
+
+-   Any change to the (non-experimental) CLI interface
+
+-   A non-trivial internal refactoring that's likely to introduce some
```

Comment:
By their nature, I don't think it's feasible to make refactorings experimental. Because that would mean you have to keep the old and new code around, which goes against the idea of a refactoring. Also, refactorings shouldn't break anything by definition.

Reply:
> shouldn’t

That’s where the devil lies :stuck_out_tongue: 

I guess it depends how you define “refactoring”. Something I had in mind in particular here is changes to the daemon protocol (say adding a new message type to make some specific operation faster), which _shouldn’t_ break anything, but happen to do so every once in a while (and for which we need to keep both versions anyways for backwards compatibility). Or just a re-implementation of a specific piece of functionality that’s well-contained (so cheap to duplicate for a short amount of time), but might be hiding some subtle bug